core/src/test/java/org/elasticsearch/index/query/HasChildQueryBuilderTests.java

```diff
@@ -0,0 +1,215 @@
+            ChildrenQuery lpq = (ChildrenQuery) query;
+            assertEquals(queryBuilder.minChildren(), lpq.getMinChildren());
+            assertEquals(queryBuilder.maxChildren(), lpq.getMaxChildren());
+            assertEquals(queryBuilder.scoreType(), lpq.getScoreType());
+            assertEquals(queryBuilder.shortCircuitCutoff(), lpq.getShortCircuitParentDocSet());
+        }
+        if (queryBuilder.innerHit() != null) {
+            assertNotNull(SearchContext.current());
+            if (query != null) {
+                assertNotNull(SearchContext.current().innerHits());
+                assertEquals(1, SearchContext.current().innerHits().getInnerHits().size());
+                assertTrue(SearchContext.current().innerHits().getInnerHits().containsKey("inner_hits_name"));
+                InnerHitsContext.BaseInnerHits innerHits = SearchContext.current().innerHits().getInnerHits().get("inner_hits_name");
+                assertEquals(innerHits.size(), 100);
+                assertEquals(innerHits.sort().getSort().length, 1);
+                assertEquals(innerHits.sort().getSort()[0].getField(), STRING_FIELD_NAME);
+            } else {
+                assertNull(SearchContext.current().innerHits());
+            }
+        }
+    }
+
+    public void testIllegalValues() {
+        QueryBuilder query = RandomQueryBuilder.createQuery(random());
+        try {
+            new HasChildQueryBuilder(null, query);
+            fail("must not be null");
+        } catch (IllegalArgumentException ex) {
+
+        }
+
+        try {
+            new HasChildQueryBuilder("foo", null);
+            fail("must not be null");
+        } catch (IllegalArgumentException ex) {
+
+        }
+        HasChildQueryBuilder foo = new HasChildQueryBuilder("foo", query);// all good
+        try {
+            foo.scoreType(null);
+            fail("must not be null");
+        } catch (IllegalArgumentException ex) {
+
+        }
+        final int positiveValue = randomIntBetween(0, Integer.MAX_VALUE);
+        try {
+            foo.shortCircuitCutoff(randomIntBetween(Integer.MIN_VALUE, -1));
+            fail("must not be negative");
+        } catch (IllegalArgumentException ex) {
+
+        }
+
+        foo.shortCircuitCutoff(positiveValue);
+        assertEquals(positiveValue, foo.shortCircuitCutoff());
+
+        try {
+            foo.minChildren(randomIntBetween(Integer.MIN_VALUE, -1));
+            fail("must not be negative");
+        } catch (IllegalArgumentException ex) {
+
+        }
+        foo.minChildren(positiveValue);
+        assertEquals(positiveValue, foo.minChildren());
+        try {
+            foo.maxChildren(randomIntBetween(Integer.MIN_VALUE, -1));
+            fail("must not be negative");
+        } catch (IllegalArgumentException ex) {
+
+        }
+
+        foo.maxChildren(positiveValue);
+        assertEquals(positiveValue, foo.maxChildren());
+    }
+
+    public void testParseFromJSON() throws IOException {
```

Comment:
do we really need this test? We already test that we can parse whatever the builder outputs as part of its toXContent output. That should be enough unless this query can be represented in a way that the builder doesn't know about and doesn't print out. makes sense?